experimental/devices/ccs811/ccs811.go

```diff
@@ -0,0 +1,361 @@
+will depend on the customer use-case and application.
+
+• For devices which are powered for >24 hours at a time:
+• During the first 500 hours – save the baseline every
+24-48 hours.
+• After the first 500 hours – save the baseline every 5-7 days.
+
+• For devices which are powered <24 hours at a time:
+• If the device is run in, save the baseline before power down
+• If multiple operating modes are used, a separate baseline should be stored for each
+• The baseline should only be restored when the resistance is stable (typically 20-30 minutes)
+• If changing from a low to high power mode (without spending at least 10 minutes in idle),
+  the sensor resistance should be allowed to settle again before restoring the baseline
+
+Note(s):
+1) If a value is written to the BASELINE register while the sensor is stabilising, the output of the TVOC and eCO2 calculations may be higher than expected.
+2) The baseline must be written after the conditioning period
+*/
+func (d *Dev) SetBaseline(baseline []byte) error {
+	w := []byte{baselineReg, baseline[0], baseline[1]}
+	if err := d.c.Tx(w, nil); err != nil {
+		return err
+	}
+	return nil
+}
+
+// SensorValues represents data read from the sensor.
+// Data are populated based on NeededData parameter.
+type SensorValues struct {
+	ECO2           int
+	VOC            int
+	Status         byte
+	ErrorID        SensorErrorID
+	RawDataCurrent int // current in uA
+	RawDataVoltage int // voltage from 0-1023, where 1023 = 1.65V
+}
+
+// Sense provides data from the sensor.
+// ReadData parameter specifies which data should be read.
+func (d *Dev) Sense(mode NeededData) (*SensorValues, error) {
+	read := make([]byte, mode)
+	err := d.c.Tx([]byte{algoResultsReg}, read)
+	if err != nil {
+		return nil, err
+	}
+	sv := &SensorValues{}
+	if mode >= ReadCO2 {
+		// exptected range: 400ppm to 8192ppm
+		// 0x3F is used to erase randomly set top bits causing value out of range given by specs
+		sv.ECO2 = int(uint32(read[0]&0x3F)<<8 | uint32(read[1]))
+	}
+	if mode >= ReadCO2VOC {
+		// expected range: 0ppb to 1187ppb
+		// 0x7 is used to erase randomly set top bits causing value out of range given by specs
+		sv.VOC = int(uint32(read[2]&0x7)<<8 | uint32(read[3]))
+	}
+	if mode >= ReadCO2VOCStatus {
+		sv.Status = read[4]
+	}
+	if mode == ReadAll {
+		sv.ErrorID = SensorErrorID(read[5])
+		sv.RawDataCurrent, sv.RawDataVoltage = valuesFromRawData(read[6:])
+	}
+
+	return sv, nil
+}
+
+// parse current and voltage from raw data
+func valuesFromRawData(data []byte) (current int, voltage int) {
+	current = int(data[0] >> 2)
+	voltage = int((uint16(data[0]&0x03) << 8) | uint16(data[1]))
+	return current, voltage
+}
+
+type FwVersions struct {
+	HWIdentifier       byte
+	HWVersion          byte
+	BootVersion        string
+	ApplicationVersion string
+}
+
+func (d *Dev) GetFirmwareData() (version *FwVersions, err error) {
+	version = &FwVersions{}
+	hwid := make([]byte, 1)
```

Comment:
~~You don't need the next 3 transactions; do one starting at 0x20 reading 5 bytes and use these.~~

(me reads the datasheet)

Wow this device is insane. Maybe even worth a comment specifically to note that while this device _looks_ like a memory mapped registers, it's not.

Reply:
I'm not very good in sensors internals, so if you have some specific idea (paragraph) what to put there, let me know, or add it anytime later